Documentation/UPI/libvirt/libvirt.md

```diff
@@ -0,0 +1,148 @@
+controlPlane:
+  hyperthreading: Enabled   
+  name: master 
+  replicas: 3 
+metadata:
+  name: CLUSTER_NAME
+networking:
+  clusterNetwork:
+  - cidr: 10.100.0.0/14 
+    hostPrefix: 23 
+  networkType: OpenShiftSDN
+  serviceNetwork: 
+  - 172.30.0.0/16
+platform:
+  none: {} 
+pullSecret: 'SECRET FROM RED HAT CLUSTER MANAGER'
+sshKey: 'YOUR SSH PUBLIC KEY' 
+```
+
+Modify it accordingly to the size and the configuration of your cluster and then remember to backup it, because the OpenShift Installer will remove it after generating the Ignition configuration files.
+
+
+### Generate the Ignition configuration files
+```
+$ openshift-install create ignition-configs
+```
+
+### Provision the VMs
+A simple high-available OKD cluster would include 3 control planes and 2 workers. So for such configuration you need for 5 VMs + 1 for boostrapping.
+Configure the VM for booting them with the kernel and initramfs you previously downloaded, with the following parameters:
+- KERNEL: `/path/of/fedora-coreos-31.20191217.2.0-installer-kernel-x86_64`;
+- INITRAMFS: `/path/of/fedora-coreos-31.20191217.2.0-installer-initramfs.x86_64.img`;
+- BOOT ARGS: `ip=dhcp rd.neednet=1 coreos.inst=yes coreos.inst.install_dev=vda coreos.inst.image_url=http://WEB_SERVER_IP:PORT/fedora-coreos-31.20191217.2.0-metal.x86_64.raw.xz coreos.inst.ignition_url=http://WEB_SERVER_IP:PORT/SERVER_ROLE.ign`.
+Replace `SERVER_ROLE` with `bootstrap` for the bootstrap server, with `master` for control plane servers, and with `worker` for worker servers.
+Replace `WEB_SERVER_IP:PORT` with the endpoint of the webserver you previously configured.
+
+Here how you can configure Direct Kernel Boot in `virt-manager`:
+![Direct kernel boot](Images/direct_kernel_boot.png)
+
+**NOTE:** CoreOS installer automatically reboot the server after installation, so when the installation of Fedora CoreOS is finished shut down the VMs, otherwise the installer will keep installing the OS.
+
+After the installation disable the Direct Kernel Boot option, as we don't need it anymore.
+
+At this point you can add the VMs MAC address to the network configuration in libvirt, just like the example [above](#configure-libvirt-network). Stop and start the network after that:
+```
+virsh# net-destroy openshift
+virsh# net-start openshift
+```
+
+### Start the bootstrap server
+After every server in your cluster was provisioned, start the bootstrap server.
+By default, Fedora CoreOS will install the OS from the official OSTree image, so we have to wait a few minutes for the machine-config-daemon to pull and install the pivot image from quay.io. This image is necessary for the kubelet service, as the official Fedora CoreOS image does not include hyperkube.
+After the image was pulled and installed, the server will be rebooted by itself.
+When the server is up again wait for the API service and the MachineConfig service to be spawned (check for the ports 6443 and 22623). Check also for the status of the `bootkube.service`.
```

Comment:
Some more services to watch:
* `machine-config-daemon-pull` pulls MCD image and extracts binary to `/usr/local/bin/machine-config-daemon`
* `machine-config-daemon-first-boot` updates FCOS to latest FCOS with CRI-O
* `crio-configure` sets up CRI-O config
* `crio` would run CRI-O daemon on bootstrap node so that it could be used by kubelet
* `kubelet` would run a temporary kubelet
* output of `podman ps -a / podman logs / podman inspect` and `crictl ps -a / crictl logs / crictl inspect` to find out details about podman/crio containers

Reply:
I think checking for such services is too much low-level for the average user that simply wants to bootstrap an OKD cluster.
Yours are really useful suggestions, @vrutkovs, but I think it would be better for they to be mentioned and properly explained in a boostrap troubleshooting-related guide. What do you think?

I wrote about checking for the ports, because they are the simple basic thing that anyone could check for, while waiting for the cluster to be bootstrapped.

Reply:
Yeah, perhaps these kind of details should be mentioned in detailed guide. https://github.com/openshift/installer/blob/master/docs/user/troubleshooting.md is probably a good starting point

Reply:
OK, then, if you agree, I'd resolve this conversation as improving the troubleshooting guide should be considered a different PR.